nixos/modules/security/vault/keys.nix

```diff
@@ -0,0 +1,420 @@
+      };
+
+      template = mkOption {
+        default = null;
+        type = types.nullOr types.lines;
+        description = "Specifies a custom HCL template to render the key. Overrides built-in sink renderers.";
+      };
+
+      backends = {
+        kv = mkOption {
+          default = null;
+          type = types.nullOr backendTypes.kv;
+          description = "Configuration for the KeyValues backend.";
+        };
+
+        pki = mkOption {
+          default = null;
+          type = types.nullOr backendTypes.pki;
+          description = "Configuration for the PKI backend.";
+        };
+      };
+
+      backendName = mkOption {
+        default = null;
+        example = "pki_int";
+        type = types.nullOr types.str;
+        description = "The name of the Vault backend. Default is to use the default name for the backend type.";
+      };
+
+      # TODO: We need to get vault-agent to reload whenever this changes, how?
+      sinks = mkOption {
+        default = { "${name}" = { }; };
+        type = types.attrsOf sinkType;
+        description = "List of sink configurations that describe how to render this key to a file. If not set, the key name will be used.";
+      };
+
+      postRenew = {
+        command = mkOption {
+          default = "";
+          type = types.str;
+          description = "Optional command to run after the key is renewed.";
+        };
+
+        restart = mkOption {
+          default = false;
+          type = types.bool;
+          description = "Whether to automatically restart or reload systemd units when the key is renewed.";
+        };
+      };
+
+      dependency = {
+        units = mkOption {
+          default = [ ];
+          example = "[ \"nginx\" ]";
+          type = types.listOf types.str;
+          description = "Dependent systemd units for the key.";
+        };
+      };
+
+      external = mkOption {
+        default = false;
+        type = types.bool;
+        description = "Specifies that this key will be pushed externally by a tool rather than being pulled by the host's agent.";
+      };
+    };
+  });
+
+  # build all the vault agent templates for a key
+  buildAgentTemplates = key:
+    let
+      defaults = {
+        createDestDirs = false;
+        sandboxPath = key.folder;
+      };
+    in
+    (forEach key.templates (template: ({
+      sourceFile = pkgs.writeText "vault-key-${template.id}.ctmpl" template.text;
+      destFile = "${key.folder}/.${template.id}.tmp";
+    } // defaults)));
+
+  # builds the final set of keys
+  finalKeys = (flip mapAttrs config.security.vault-keys (_: key: key // { templates = createConsulTemplates key; }));
+in
+{
+  options = {
+    security.vault-keys = mkOption {
+      default = { };
+      type = types.attrsOf keyType;
+      description = "Attributes of vault keys to deploy.";
+    };
+
+    reflection.vault-keys = mkOption {
+      default = finalKeys;
+      type = types.attrs;
```

Comment:
`types.attrs` is deprecated: `types.attrsOf types.anything` is the replacement, I believe.
If the value is a script, we could do `types.attrsOf (types.oneOf [ types.package types.path ])` I think?